synapse/logging/context.py

```diff
@@ -287,6 +288,46 @@ def __str__(self) -> str:
             return str(self.request)
         return "%s@%x" % (self.name, id(self))
 
+    @classmethod
+    def current_context(cls) -> LoggingContextOrSentinel:
+        """Get the current logging context from thread local storage
+
+        This exists for backwards compatibility. ``current_context()`` should be
+        called directly.
+
+        Returns:
+            LoggingContext: the current logging context
+        """
+        warnings.warn(
+            "synapse.logging.context.LoggingContext.current_context() is deprecated "
+            "in favor of synapse.logging.context.current_context().",
+            PendingDeprecationWarning,
```

Comment:
Out of curiosity, why `PendingDeprecationWarning` rather than `DeprecationWarning`?

Reply:
I'm not sure it really matters in our case? The definitions from [the docs](https://docs.python.org/3/library/warnings.html):

> [`DeprecationWarning`](https://docs.python.org/3/library/exceptions.html#DeprecationWarning): Base category for warnings about deprecated features when those warnings are intended for other Python developers (ignored by default, unless triggered by code in `__main__`).
>
> [`PendingDeprecationWarning`](https://docs.python.org/3/library/exceptions.html#PendingDeprecationWarning): Base category for warnings about features that will be deprecated in the future (ignored by default).

Although further reading of the docs for `PendingDeprecationWarning` do say:
> Base class for warnings about features which are obsolete and expected to be deprecated in the future, but are not deprecated at the moment.
>
> This class is rarely used as emitting a warning about a possible upcoming deprecation is unusual, and `DeprecationWarning` is preferred for already active deprecations.

I'll change it to `DeprecationWarning`.